### PR TITLE
[metricbeat] [gcp] remove compute metadata cache

### LINKED ISSUE #32734: GCP metadata instance cache is never refreshed but also never used
`gcp.compute` metricset implements a caching mechanism for GCP Instance metadata. These metadata are collected as part of instance data collection.

An investigation into how the current caching works highlighted a major flaw in it's invalidation behaviour.

**TL;DR**: current cache behaviour is useless and cache is not used, the reason for having a cache (high number of calls to `instances.Get` GCP API) has been removed and invalidating the cache is problematic (there is no way to know when data changes in GCP, with the high risk of providing stale data). Due to this the proposal is to remove the code and introduce it back if needed, as we foresee this not to be needed anymore.

we reduced a lot the number of API calls involved that justified caching and caching invalidation for this use case is very error prone; can we remove it entirely?

## How instance metadata collection works

<details>
<summary>Details</summary>

Instance metadata collection is enabled when `exclude_labels` is set to `true` in metricset configuration.

When enabled, the code instantiate a [`gcp.MetadataService`](https://github.com/elastic/beats/blob/7449e5c4b944c661299de8099d5423bafd458ee2/x-pack/metricbeat/module/gcp/metadata.go#L18):
https://github.com/elastic/beats/blob/7449e5c4b944c661299de8099d5423bafd458ee2/x-pack/metricbeat/module/gcp/metrics/metricset.go#L215-L219

which in the case of `gcp.compute` means calling this function:
https://github.com/elastic/beats/blob/7449e5c4b944c661299de8099d5423bafd458ee2/x-pack/metricbeat/module/gcp/metrics/compute/metadata.go#L29-L38

For each collection period, the `gcp.MetadataService` `Metadata` function is called to gather instance metadata.

We can assume, as knowledge of why this cache is in place has been lost, that the reason was avoiding calling the [`instances.get` GCP API endpoint](https://cloud.google.com/compute/docs/reference/rest/v1/instances/get) at each collection iteration, to avoid costly API calls and network latency. This API was called for each instance to retrieve metadata for, which also poses a scaling problem with big instance number.

This behaviour has been altered in March 2022 with this PR https://github.com/elastic/beats/pull/30154, where the single API call per instance has been improved by using [`instances.aggregatedList` GCP API](https://cloud.google.com/compute/docs/reference/rest/v1/instances#aggregatedlist) that returns batched results (up to 500 instances per result page).

So with the current implementation concerns for scaling (performance and cost wise) for this operation are greatly reduced; with 1000 instances monitored with metadata collection enabled metricbeat now performs 2 API calls instead of 1000.

</details>

## The cache

<details>
<summary>Details</summary>

The cache, called `instanceCache` in `compute.NewMetadataService` is instantiated with:
https://github.com/elastic/beats/blob/3fda4699146f9e2aa0b472a45302dc234ce53b44/x-pack/metricbeat/module/gcp/metrics/compute/metadata.go#L35

Where values are:
https://github.com/elastic/beats/blob/3fda4699146f9e2aa0b472a45302dc234ce53b44/x-pack/metricbeat/module/gcp/metrics/compute/metadata.go#L24-L25

[`common.Cache`](https://github.com/elastic/beats/blob/fc2212d7056108f0cc147791554e5f2f6099db4d/libbeat/common/cache.go) is our libbeat common cache implementation.

</details>

## What is cached

<details>
<summary>Details</summary>

When `gcp.MetadataService` `Metadata` is called its first step is to gather current instance metadata (by instance ID), which triggers a cache refresh:
https://github.com/elastic/beats/blob/7449e5c4b944c661299de8099d5423bafd458ee2/x-pack/metricbeat/module/gcp/metrics/compute/metadata.go#L144-L145

Iterating over pages of results cache refresh add the result of `instances.aggregatedList` API to `instanceCache`.

</details>

### Issue 1 - expired elements are not cleaned up

TL;DR: once filled the cache is never really refreshed because the condition expects an empty cache, but this case does not happen.

<details>
<summary>Details</summary>

The problem starts with this lines:
https://github.com/elastic/beats/blob/7449e5c4b944c661299de8099d5423bafd458ee2/x-pack/metricbeat/module/gcp/metrics/compute/metadata.go#L173-L176

This check was implemented to prevent refreshing the cache at each iteration, and wait for cache to be empty before performing a cache refresh.

But in which case `Size()` return `0`? Looking at its implementation:
https://github.com/elastic/beats/blob/fc2212d7056108f0cc147791554e5f2f6099db4d/libbeat/common/cache.go#L230-L236

we notice that `Size()` returns the number of all active and **expired** elements (that have not been yet cleaned up).

When does clean-up happens? When [`Cleanup()`](https://github.com/elastic/beats/blob/fc2212d7056108f0cc147791554e5f2f6099db4d/libbeat/common/cache.go#L197-L201) is  called, something that **never happens** in metricbeat code.

This is the first issue: once filled the cache is never really refreshed because the condition expects an empty cache, but this case does not happen.

</details>

### Issue 2 - element expiration time is extended on access

TL;DR: the expiration time for elements is changed on access, so elements would never be cleaned up (and cache size would never be 0 after the first fill).

<details>
<summary>Details</summary>

But there is a second problem: even if the cache was correctly cleaned-up, when an element is considered expired?   
The `get` function implementing cached element retrieval may update the element expiration time, depending on an internal attribute (`accessExpire`). 

Metricbeat code instantiate the cache with `common.NewCache(...)` which implemented is:
https://github.com/elastic/beats/blob/fc2212d7056108f0cc147791554e5f2f6099db4d/libbeat/common/cache.go#L82-L84

The second parameter is `accessExpire`:
https://github.com/elastic/beats/blob/fc2212d7056108f0cc147791554e5f2f6099db4d/libbeat/common/cache.go#L102

which is set to `true` by default. 

This is the second problem: the expiration time for elements is changed on access, so elements would never be cleaned up (and cache size would never be 0 after the first fill).

</details>

### Issue 3 - cache expires between periods

TL;DR: all elements expire after 30 seconds and we access them every 60 seconds, but due to not cleaning them up they are not retrieved from cache and the cache is not refreshed neither.

<details>
<summary>Details</summary>

Up to this point it seems the cache is not really working as expected, but there is another wrong behaviour.

[Cache timeout](https://github.com/elastic/beats/blob/fc2212d7056108f0cc147791554e5f2f6099db4d/libbeat/common/cache.go#L71) is initialized to `cacheTTL`, which is set to:
https://github.com/elastic/beats/blob/7449e5c4b944c661299de8099d5423bafd458ee2/x-pack/metricbeat/module/gcp/metrics/compute/metadata.go#L24

In `gcp` any period less that `60` seconds cannot be configured because that's the minimum valid for collection on GCP API side. 

This is the third issue: all elements expire after 30 seconds and we access them every 60 seconds, but due to not cleaning them up they are not retrieved from cache and the cache is not refreshed neither.

</details>

### Issue 4 - MetadataService is instantiated at each period

The cache is instantiated in [`compute.NewMetadataService`](https://github.com/elastic/beats/blob/3fda4699146f9e2aa0b472a45302dc234ce53b44/x-pack/metricbeat/module/gcp/metrics/compute/metadata.go#L29). This function is:
- called by [`gcp.NewMetadataService`](https://github.com/elastic/beats/blob/3fda4699146f9e2aa0b472a45302dc234ce53b44/x-pack/metricbeat/module/gcp/metrics/metadata_services.go#L17)
- which is called by [`gcp.eventMapping`](https://github.com/elastic/beats/blob/3fda4699146f9e2aa0b472a45302dc234ce53b44/x-pack/metricbeat/module/gcp/metrics/metricset.go#L216)
- which is called by [`gcp.Fetch`](https://github.com/elastic/beats/blob/3fda4699146f9e2aa0b472a45302dc234ce53b44/x-pack/metricbeat/module/gcp/metrics/metricset.go#L194)

This is the fourth issue: the cache instance is created for each data collection, so it always starts empty.

## Conclusions

Due to these 4 issues:
- the cache is effectively not used
- it would be filled and never refreshed
- expired items would never been cleaned up
- items would never expire because their expiration time is expanded upon access
- the cache is instantiated fresh at each collection period

the cache behaviour is wrong **and** the cache is never used.

### PR DESCRIPTION
**Pull request by ghost**
## What does this PR do?

Removes compute metadata cache.

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

- Closes: #32734

### CODE REVIEW

**Comment by endorama** [comment]
@gpop63 I like the direction this is headed, can we move forward with it?

File: x-pack/metricbeat/module/gcp/metrics/compute/metadata.go [hunk 1]
```diff
@@ -9,32 +9,33 @@ import (
...
 	"google.golang.org/api/option"
 	monitoringpb "google.golang.org/genproto/googleapis/monitoring/v3"
 
-	"github.com/elastic/beats/v7/libbeat/common"
 	"github.com/elastic/beats/v7/x-pack/metricbeat/module/gcp"
 	"github.com/elastic/elastic-agent-libs/logp"
 )
 
+/*
 const (
```
Comment (kaiyan-sheng) [inline x-pack/metricbeat/module/gcp/metrics/compute/metadata.go:22]: can we remove these const since we are not using it anymore? :)

**Review by endorama — CHANGES_REQUESTED** [review]
Logic is correct, can you please clean up all the commented code?

File: x-pack/metricbeat/module/gcp/metrics/compute/metadata_test.go [hunk 2]
```diff
@@ -67,8 +64,8 @@ var fake = &monitoring.TimeSeries{
 }
 
 var m = &metadataCollector{
-	projectID:     "projectID",
-	instanceCache: common.NewCache(30*time.Second, 13),
+	projectID: "projectID",
+	// instanceCache: common.NewCache(30*time.Second, 13),
```
Comment (endorama) [inline x-pack/metricbeat/module/gcp/metrics/compute/metadata_test.go:68]: Can we get rid of this too?

File: x-pack/metricbeat/module/gcp/metrics/compute/metadata.go [hunk 3]
```diff
@@ -9,59 +9,50 @@ import (
...
-	regions       []string
-	opt           []option.ClientOption
-	instanceCache *common.Cache
-	logger        *logp.Logger
+	projectID string
+	zone      string
+	region    string
+	regions   []string
+	opt       []option.ClientOption
+	// instanceCache *common.Cache
```
Comment (endorama) [inline x-pack/metricbeat/module/gcp/metrics/compute/metadata.go:53]: Can we get rid of this too?

**Review by endorama — APPROVED** [review]

**Review by kaiyan-sheng — APPROVED** [review]
